### PR TITLE
plugins.vk: add support for WAF cookie

### LINKED ISSUE #4637: plugins.vk: fixes required
### Checklist

- [X] This is a plugin issue and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest build from the master branch

### Description

Needs support for the web application firewall cookie.

~Seems necessary to add all parameters to the API POST data in `_get_streams()` now.~

ref: https://github.com/streamlink/streamlink/pull/4613#issuecomment-1173040359

### Debug log

```text
$ streamlink -l debug https://vk.com/video-211154316_456239169
[cli][debug] OS:         Linux-4.9.0-18-amd64-x86_64-with-debian-9.13
[cli][debug] Python:     3.7.3
[cli][debug] Streamlink: 4.1.0+45.gb277114d
[cli][debug] Dependencies:
[cli][debug]  isodate: 0.6.1
[cli][debug]  lxml: 4.9.1
[cli][debug]  pycountry: 22.3.5
[cli][debug]  pycryptodome: 3.15.0
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.28.1
[cli][debug]  websocket-client: 1.3.3
[cli][debug]  importlib-metadata: 4.12.0
[cli][debug] Arguments:
[cli][debug]  url=https://vk.com/video-211154316_456239169
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin vk for URL https://vk.com/video-211154316_456239169
[plugins.vk][debug] video ID: -211154316_456239169
[plugins.vk][error] Could not parse API response
error: No playable streams found on this URL: https://vk.com/video-211154316_456239169
```

### PR DESCRIPTION
**Pull request by mkbloke**
Also corrected the API POST `params`/`data` dicts.

closes #4637

### CODE REVIEW

File: src/streamlink/plugins/vk.py [hunk 1]
```diff
@@ -25,6 +26,34 @@
...
+
+    def get_and_save_new_waf_cookie(self):
+        log.info("Getting new cookie.")
+        res = self.session.http.get("https://vk.com/")
+        cookie = self.session.http.cookies.get(self.HASH_COOKIE)
+        if not cookie:
+            log.error("Failed to get expected hash cookie.")
+            return
+        key = md5(cookie.encode("utf-8")).hexdigest()
+        res = self.session.http.get(f"{res.url}&key={key}")
```
Comment (bastimeyer) [inline src/streamlink/plugins/vk.py:47]: The `res` var is unused.

Setting the `key` parameter on the `params` keyword is also better than modifying the request URL.

Also, if you want to make mypy happy, then return False where you currently return None (the linting rule is currently disabled, hence why there's no failure).

**Review by bastimeyer — COMMENTED** [review]
The first URL from the tests is returning this:

> ```
> $ streamlink 'https://vk.com/video-9944999_456239622' best -l debug
> ...
> [plugins.vk][debug] Saved cookie not found.
> [plugins.vk][info] Getting new cookie.
> error: Unable to open URL: https://vk.com/429.html?hash429=wMF4IzvUyM5acpz9RUB7S4wasLQWDuHjX7ptYhUSQedbwJd8ViiIDIUF4MapFBvaDBhxzkauoVfsYbzXN5VPlqMwIaoPMD2aWbxtyC4GCp1JMgkjuRc6&redirect429=/&key=87477569783524fc736506c9a64450f4 (429 Client Error: Too Many Requests for url: https://vk.com/)
> ```

**Review by bastimeyer — CHANGES_REQUESTED** [review]
Don't add hooks to the entire HTTP session. This is bad.

That should do the trick:

```diff
diff --git a/src/streamlink/plugins/vk.py b/src/streamlink/plugins/vk.py
index 33a6365c..c72e9458 100644
--- a/src/streamlink/plugins/vk.py
+++ b/src/streamlink/plugins/vk.py
@@ -14,6 +14,7 @@ from streamlink.plugin import Plugin, PluginError, pluginmatcher
 from streamlink.plugin.api import validate
 from streamlink.stream.dash import DASHStream
 from streamlink.stream.hls import HLSStream
+from streamlink.utils.url import update_qsd
 
 log = logging.getLogger(__name__)
 
@@ -29,24 +30,25 @@ class VK(Plugin):
     HASH_COOKIE = "hash429"
     SOLUTION_COOKIE = "solution429"
 
-    def check_for_waf(self, res, **kwargs):
-        if res.headers.get("X-WAF-Backend-Status") == "challenge_success" and hasattr(self, "original_request"):
-            self.session.http.cookies.update(res.cookies)
-            self.session.http.cookies.pop(self.HASH_COOKIE, None)
-            self.original_request.headers.pop("Cookie", None)
-            self.original_request.prepare_cookies(self.session.http.cookies)
-            res = self.session.http.send(self.original_request)
-            del self.original_request
-            self.save_cookies(cookie_filter=lambda c: c.name == self.SOLUTION_COOKIE)
-            return res
-
-        if res.status_code == 302 and res.cookies.get(self.HASH_COOKIE):
-            log.debug("Getting new WAF cookie")
-            cookie = res.cookies.get(self.HASH_COOKIE)
-            key = md5(cookie.encode("utf-8")).hexdigest()
-            res.headers["Location"] = f"{res.headers['Location']}&key={key}"
-            self.original_request = res.request
-            return res
+    def _get_cookies(self):
+        if self.session.http.cookies.get(self.SOLUTION_COOKIE, domain=".vk.com"):
+            log.debug("Using cached WAF cookie")
+            return
+
+        def on_response(res, **kwargs):
+            if res.headers.get("x-waf-redirect") == "1":
+                if not res.headers.get("X-WAF-Backend-Status"):
+                    log.debug("Getting new WAF cookie")
+                    cookie = res.cookies.get(self.HASH_COOKIE)
+                    key = md5(cookie.encode("utf-8")).hexdigest()
+                    res.headers["Location"] = update_qsd(res.headers["Location"], qsd={"key": key})
+                    return res
+                elif res.headers.get("X-WAF-Backend-Status") == "challenge_success":
+                    self.session.http.cookies.update(res.cookies)
+                    self.save_cookies(cookie_filter=lambda c: c.name == self.SOLUTION_COOKIE)
+                    return res
+
+        self.session.http.get("https://vk.com/", hooks={"response": on_response})
 
     def _has_video_id(self):
         return any(m for m in self.matches[:-1])
@@ -78,8 +80,7 @@ class VK(Plugin):
         raise NoStreamsError(self.url)
 
     def _get_streams(self):
-        self.session.http.hooks["response"] = self.check_for_waf
-
+        self._get_cookies()
         self.follow_vk_redirect()
 
         video_id = self.match.group("video_id")
```

**Comment by mkbloke** [comment]
One of the reasons for some of the logic I put into the original code is that it didn't require an initial probe.  Some of the other logic was to handle the case that if you change IP address (or maybe region, who knows, but the latter can happen if you're a VPN user), an existing `solution429` cookie becomes invalid and will need replacing.  This scenario now fails with the code above.

**Comment by bastimeyer** [comment]
Just make the initial request every time, and don't store the cookie, as it's unnecessary.

**Comment by bastimeyer** [comment]
The `self.session.http.cookies.update(res.cookies)` on `res.headers.get("X-WAF-Backend-Status") == "challenge_success"` should be redundant, as `requests` will update the session cookies after the hook has been executed anyway. Unfortunately, I don't get the cookie requirement right now, so I can't check. Doesn't matter though, going to merge.

**Pull request merged**